<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><link rel="canonical" href="https://docs.nupack.org/latest/model/" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Model Specification - NUPACK 4 User Guide</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="../extra.css" rel="stylesheet" />
        <link href="../css/version-select.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Model Specification";
        var mkdocs_page_input_path = "model.md";
        var mkdocs_page_url = "/latest/model/";
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> NUPACK 4 User Guide
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Overview</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../definitions/">Definitions</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="#">Model Specification</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#specify-a-physical-model">Specify a physical model</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#model-options">Model options</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#material">Material</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#stacking">Stacking</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#temperature">Temperature</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#salts">Salts</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#historical-options">Historical options</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#compute-loop-free-energy">Compute loop free energy</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#compute-stacking-state-free-energies">Compute stacking state free energies</a>
    </li>
    </ul>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../analysis/">Analysis Jobs</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../design/">Design Jobs</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../utilities/">Utilities Jobs</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../start/">Getting Started</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../advanced/">Advanced</a>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">NUPACK 4 User Guide</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">Model Specification</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="model-specification">Model Specification<a class="headerlink" href="#model-specification" title="Permanent link">&para;</a></h1>
<hr>

<h2 id="specify-a-physical-model">Specify a physical model<a class="headerlink" href="#specify-a-physical-model" title="Permanent link">&para;</a></h2>
<p>NUPACK 4 analysis and design jobs are run based on a physical model created using the <code>Model</code> class:</p>
<div class="highlight"><pre><span></span><code><span class="n">model1</span> <span class="o">=</span> <span class="n">Model</span><span class="p">(</span><span class="n">material</span><span class="o">=</span><span class="s1">&#39;rna&#39;</span><span class="p">,</span> <span class="n">ensemble</span><span class="o">=</span><span class="s1">&#39;stacking&#39;</span><span class="p">,</span> <span class="n">celsius</span><span class="o">=</span><span class="mi">37</span><span class="p">,</span>
    <span class="n">sodium</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">magnesium</span><span class="o">=</span><span class="mf">0.0</span><span class="p">)</span>
</code></pre></div>
<p>Any unspecified properties take on their default values (which happen to be the ones specified for <code>model1</code> above).</p>
<hr>

<h2 id="model-options">Model options<a class="headerlink" href="#model-options" title="Permanent link">&para;</a></h2>
<p>The valid options for each property are described below.</p>
<h3 id="material">Material<a class="headerlink" href="#material" title="Permanent link">&para;</a></h3>
<p>NUPACK 4.1 algorithms use the following temperature-dependent free energy parameter sets for single-material jobs (RNA, DNA, or 2’OMe-RNA), mixed-material jobs (RNA/DNA or RNA/2’OMe-RNA), or custom-material jobs, specified by the keyword <code>material</code> (default: <code>material='rna'</code>):</p>
<ul>
<li>RNA single-material parameter sets: <ul>
<li><code>rna</code> Shorthand for <code>rna06</code>.</li>
<li><code>rna06</code> Based on [<a class="citation" href="#ref-Mathews99" id="cite-Mathews99">Mathews99</a>] and [<a class="citation" href="#ref-Lu06" id="cite-Lu06">Lu06</a>] with additional parameters [<a class="citation" href="#ref-Xia98" id="cite-Xia98">Xia98</a>,<a class="citation" href="#ref-Zuker03" id="cite-Zuker03">Zuker03</a>] including coaxial stacking [<a class="citation" href="#ref-Mathews99" id="cite-Mathews99">Mathews99</a>,<a class="citation" href="#ref-Turner10" id="cite-Turner10">Turner10</a>] and dangle stacking [<a class="citation" href="#ref-Serra95" id="cite-Serra95">Serra95</a>,<a class="citation" href="#ref-Zuker03" id="cite-Zuker03">Zuker03</a>,<a class="citation" href="#ref-Turner10" id="cite-Turner10">Turner10</a>] in a user specified concentration of Na<span class="arithmatex"><span class="MathJax_Preview">^+</span><script type="math/tex">^+</script></span>.</li>
<li><code>rna95</code> Based on [<a class="citation" href="#ref-Serra95" id="cite-Serra95">Serra95</a>] with additional parameters [<a class="citation" href="#ref-Zuker03" id="cite-Zuker03">Zuker03</a>] including coaxial stacking [<a class="citation" href="#ref-Mathews99" id="cite-Mathews99">Mathews99</a>,<a class="citation" href="#ref-Turner10" id="cite-Turner10">Turner10</a>] and dangle stacking [<a class="citation" href="#ref-Serra95" id="cite-Serra95">Serra95</a>,<a class="citation" href="#ref-Zuker03" id="cite-Zuker03">Zuker03</a>,<a class="citation" href="#ref-Turner10" id="cite-Turner10">Turner10</a>] in 1.0 M Na<span class="arithmatex"><span class="MathJax_Preview">^+</span><script type="math/tex">^+</script></span>.</li>
</ul>
</li>
<li>DNA single-material parameter sets: <ul>
<li><code>dna</code> Shorthand for <code>dna04.2</code>.</li>
<li><code>dna04.2</code> Updated GT internal mismatch values [<a class="citation" href="#ref-Allawi97" id="cite-Allawi97">Allawi97</a>,<a class="citation" href="#ref-Allawi98a" id="cite-Allawi98a">Allawi98a</a>,<a class="citation" href="#ref-Allawi98b" id="cite-Allawi98b">Allawi98b</a>,<a class="citation" href="#ref-Allawi98c" id="cite-Allawi98c">Allawi98c</a>,<a class="citation" href="#ref-Peyret99" id="cite-Peyret99">Peyret99</a>,<a class="citation" href="#ref-SantaLucia04" id="cite-SantaLucia04">SantaLucia04</a>], internal asymmetry values [<a class="citation" href="#ref-SantaLucia04" id="cite-SantaLucia04">SantaLucia04</a>], and terminal mismatch values [<a class="citation" href="#ref-Turner10" id="cite-Turner10">Turner10</a>,<a class="citation" href="#ref-Mittal24" id="cite-Mittal24">Mittal24</a>]; in user-specified concentrations of Na<span class="arithmatex"><span class="MathJax_Preview">^+</span><script type="math/tex">^+</script></span>, K<span class="arithmatex"><span class="MathJax_Preview">^+</span><script type="math/tex">^+</script></span>, NH<span class="arithmatex"><span class="MathJax_Preview">^+_4</span><script type="math/tex">^+_4</script></span> and Mg<span class="arithmatex"><span class="MathJax_Preview">^{++}</span><script type="math/tex">^{++}</script></span> [<a class="citation" href="#ref-SantaLucia98" id="cite-SantaLucia98">SantaLucia98</a>,<a class="citation" href="#ref-Peyret00" id="cite-Peyret00">Peyret00</a>,<a class="citation" href="#ref-SantaLucia04" id="cite-SantaLucia04">SantaLucia04</a>].</li>
<li><code>dna04.1</code> Based on [<a class="citation" href="#ref-SantaLucia98" id="cite-SantaLucia98">SantaLucia98</a>] and [<a class="citation" href="#ref-SantaLucia04" id="cite-SantaLucia04">SantaLucia04</a>] with additional parameters [<a class="citation" href="#ref-Zuker03" id="cite-Zuker03">Zuker03</a>] including coaxial stacking [<a class="citation" href="#ref-Peyret00" id="cite-Peyret00">Peyret00</a>] and dangle stacking [<a class="citation" href="#ref-Bommarito00" id="cite-Bommarito00">Bommarito00</a>,<a class="citation" href="#ref-Zuker03" id="cite-Zuker03">Zuker03</a>] in user-specified concentrations of Na<span class="arithmatex"><span class="MathJax_Preview">^+</span><script type="math/tex">^+</script></span>, K<span class="arithmatex"><span class="MathJax_Preview">^+</span><script type="math/tex">^+</script></span>, NH<span class="arithmatex"><span class="MathJax_Preview">^+_4</span><script type="math/tex">^+_4</script></span> and Mg<span class="arithmatex"><span class="MathJax_Preview">^{++}</span><script type="math/tex">^{++}</script></span> [<a class="citation" href="#ref-SantaLucia98" id="cite-SantaLucia98">SantaLucia98</a>,<a class="citation" href="#ref-Peyret00" id="cite-Peyret00">Peyret00</a>,<a class="citation" href="#ref-SantaLucia04" id="cite-SantaLucia04">SantaLucia04</a>].</li>
</ul>
</li>
<li>2<span class="arithmatex"><span class="MathJax_Preview">'</span><script type="math/tex">'</script></span>OMe-RNA single-material parameter sets: <ul>
<li><code>merna06</code> Based on [<a class="citation" href="#ref-Kierzek06" id="cite-Kierzek06">Kierzek06</a>] using stack loop, coaxial stacking, terminal base pair, and strand association values from <code>rna_merna06</code> and all other values from <code>rna06</code>; in 0.12 M Na<span class="arithmatex"><span class="MathJax_Preview">^+</span><script type="math/tex">^+</script></span>. </li>
</ul>
</li>
<li>RNA/DNA mixed-material parameter sets:<ul>
<li><code>rna_dna06</code> Using hybrid stack loops [<a class="citation" href="#ref-Sugimoto95" id="cite-Sugimoto95">Sugimoto95</a>], hybrid internal mismatches [<a class="citation" href="#ref-Watkins11" id="cite-Watkins11">Watkins11</a>,<a class="citation" href="#ref-Sugimoto00" id="cite-Sugimoto00">Sugimoto00</a>], and chimeric stack loops [<a class="citation" href="#ref-Nakano04" id="cite-Nakano04">Nakano04</a>] with additional parameters for mixed-material loops [<a class="citation" href="#ref-Nanjundiah25" id="cite-Nanjundiah25">Nanjundiah25</a>] in a user-specified concentration of Na<span class="arithmatex"><span class="MathJax_Preview">^+</span><script type="math/tex">^+</script></span>; for use in conjunction with single-material parameter sets <code>rna06</code> and <code>dna04</code>.</li>
</ul>
</li>
<li>RNA/2<span class="arithmatex"><span class="MathJax_Preview">'</span><script type="math/tex">'</script></span>OMe-RNA mixed-material parameter sets: <ul>
<li><code>rna_merna06</code> Based on hybrid stack loops [<a class="citation" href="#ref-Kierzek06" id="cite-Kierzek06">Kierzek06</a>] with additional parameters for mixed-material loops [<a class="citation" href="#ref-Nanjundiah25" id="cite-Nanjundiah25">Nanjundiah25</a>]; for use in conjunction with single-material parameter sets <code>rna06</code> and <code>merna06</code> in 0.12 M Na<span class="arithmatex"><span class="MathJax_Preview">^+</span><script type="math/tex">^+</script></span>.</li>
</ul>
</li>
<li>Custom parameter sets: <ul>
<li><code>custom-parameters</code> Custom parameters provided in a JSON file (e.g., <code>custom-parameters.json</code>) using the same format as the provided parameter files. Provide <span class="arithmatex"><span class="MathJax_Preview">\Delta</span><script type="math/tex">\Delta</script></span>G<span class="arithmatex"><span class="MathJax_Preview">_{37}</span><script type="math/tex">_{37}</script></span>(loop) and <span class="arithmatex"><span class="MathJax_Preview">\Delta</span><script type="math/tex">\Delta</script></span>H(loop) values to allow calculations at different temperatures or only <span class="arithmatex"><span class="MathJax_Preview">\Delta</span><script type="math/tex">\Delta</script></span>G(loop) values to allow calculations at one temperature. Place the JSON file in the same directory as the default parameter files (specify <code>material = 'custom-parameters'</code>) or specify the full path to the file (<code>material = 'path/to/my/custom-parameters.json'</code>).</li>
</ul>
</li>
</ul>
<p>Free energies are expressed in kcal/mol. Base pairs are either <a href="../definitions/#watson-crick-pairs">Watson-Crick pairs</a> or <a href="../definitions/#watson-crick-pairs">wobble pairs</a>.</p>
<hr>

<h3 id="stacking">Stacking<a class="headerlink" href="#stacking" title="Permanent link">&para;</a></h3>
<p>NUPACK 4 algorithms perform calculations on the following complex ensembles specified by the keyword <code>ensemble</code> (default: <code>ensemble='stacking'</code>):</p>
<ul>
<li>
<p><code>stacking</code>
Complex ensemble with coaxial and dangle stacking (ensemble <span class="arithmatex"><span class="MathJax_Preview">\overline\Gamma^\shortparallel(\phi)</span><script type="math/tex">\overline\Gamma^\shortparallel(\phi)</script></span>).</p>
</li>
<li>
<p><code>dangle-stacking</code>
Complex ensemble with dangle stacking.</p>
</li>
<li>
<p><code>coaxial-stacking</code>
Complex ensemble with coaxial stacking.</p>
</li>
<li>
<p><code>nostacking</code>
Complex ensemble without coaxial and dangle stacking (ensemble <span class="arithmatex"><span class="MathJax_Preview">\overline\Gamma(\phi)</span><script type="math/tex">\overline\Gamma(\phi)</script></span>).</p>
</li>
</ul>
<hr>

<h3 id="temperature">Temperature<a class="headerlink" href="#temperature" title="Permanent link">&para;</a></h3>
<ul>
<li><code>celsius</code>
Temperature is specified in <span class="arithmatex"><span class="MathJax_Preview">^\circ</span><script type="math/tex">^\circ</script></span>C using the keyword <code>celsius</code> (default: <code>celsius=37</code>).</li>
<li><code>kelvin</code>
Alternatively, the temperature can be specified in K using the keyword <code>kelvin</code>.</li>
</ul>
<hr>

<h3 id="salts">Salts<a class="headerlink" href="#salts" title="Permanent link">&para;</a></h3>
<p>NUPACK 4.1 algorithms support the following salt conditions: </p>
<ul>
<li>For RNA single-material jobs using <code>rna06</code>:  <ul>
<li><code>sodium</code> Based on [<a class="citation" href="#ref-Nanjundiah25" id="cite-Nanjundiah25">Nanjundiah25</a>], the concentration of sodium ions, [Na<span class="arithmatex"><span class="MathJax_Preview">^+</span><script type="math/tex">^+</script></span>], is specified in units of molar (default: 1.0, range: [0.05,1.0]) using the keyword <code>sodium</code>.</li>
<li><code>magnesium</code> Only 0.0 M [Mg<span class="arithmatex"><span class="MathJax_Preview">^{++}</span><script type="math/tex">^{++}</script></span>] is supported. </li>
</ul>
</li>
<li>For RNA single-material jobs using <code>rna95</code>, the only supported salt conditions [<a class="citation" href="#ref-Mathews99" id="cite-Mathews99">Mathews99</a>, <a class="citation" href="#ref-Lu06" id="cite-Lu06">Lu06</a>] are 1.0 M [Na<span class="arithmatex"><span class="MathJax_Preview">^+</span><script type="math/tex">^+</script></span>]. </li>
<li>For DNA single-material jobs using <code>dna04.1</code>/<code>dna04.2</code>:<ul>
<li><code>sodium</code> Based on [<a class="citation" href="#ref-SantaLucia98" id="cite-SantaLucia98">SantaLucia98</a>,<a class="citation" href="#ref-SantaLucia04" id="cite-SantaLucia04">SantaLucia04</a>], the sum of the concentrations of (monovalent) sodium, potassium, and ammonium ions, [Na<span class="arithmatex"><span class="MathJax_Preview">^+</span><script type="math/tex">^+</script></span>]+[K<span class="arithmatex"><span class="MathJax_Preview">^+</span><script type="math/tex">^+</script></span>]+[NH<span class="arithmatex"><span class="MathJax_Preview">^+_4</span><script type="math/tex">^+_4</script></span>], is specified in units of molar (default: 1.0, range: [0.05,1.1]) using the keyword sodium.</li>
<li><code>magnesium</code> Based on [<a class="citation" href="#ref-Peyret00" id="cite-Peyret00">Peyret00</a>,<a class="citation" href="#ref-Koehler05" id="cite-Koehler05">Koehler05</a>], the concentration of (divalent) magnesium ions, [Mg<span class="arithmatex"><span class="MathJax_Preview">^{++}</span><script type="math/tex">^{++}</script></span>], is specified in units of molar (default: 0.0, range: [0.0,0.2]) using the keyword <code>magnesium</code>.</li>
</ul>
</li>
<li>For 2’OMe-RNA single-material jobs [<a class="citation" href="#ref-Kierzek06" id="cite-Kierzek06">Kierzek06</a>] using <code>merna06</code>, the only supported salt conditions are 0.12 M [Na<span class="arithmatex"><span class="MathJax_Preview">^+</span><script type="math/tex">^+</script></span>]. </li>
<li>For RNA/DNA mixed-material jobs using <code>rna_dna06</code>:  <ul>
<li><code>sodium</code> Based on [<a class="citation" href="#ref-Nanjundiah25" id="cite-Nanjundiah25">Nanjundiah25</a>], the concentration of sodium ions, [Na<span class="arithmatex"><span class="MathJax_Preview">^+</span><script type="math/tex">^+</script></span>], is specified in units of molar (default: 1.0, range: [0.12,1.0]) using the keyword <code>sodium</code>.</li>
<li><code>magnesium</code> Only 0.0 M [Mg<span class="arithmatex"><span class="MathJax_Preview">^{++}</span><script type="math/tex">^{++}</script></span>] is supported.</li>
</ul>
</li>
<li>For RNA/2’OMe-RNA mixed-material jobs [<a class="citation" href="#ref-Kierzek06" id="cite-Kierzek06">Kierzek06</a>] using <code>rna_merna06</code>, the only supported salt conditions are 0.12 M [Na<span class="arithmatex"><span class="MathJax_Preview">^+</span><script type="math/tex">^+</script></span>]. </li>
</ul>
<hr>

<!-- ### Wobble pairs

G$\cdot$U RNA wobble pairs are enabled by default in the provided RNA parameter sets.
G$\cdot$T DNA wobble pairs are disabled by default in the provided DNA parameter set.
However, wobble pairs may be manually enabled or disabled using the additional `wobble` parameter during model construction (e.g. `Model(..., wobble=True)`).
In the historical ensembles, a duplex terminated by a wobble pair is forbidden.


<hr>  -->

<div class="admonition example">
<p class="admonition-title">Examples</p>
<ul>
<li>Define a model for DNA calculations at 23 <span class="arithmatex"><span class="MathJax_Preview">^\circ</span><script type="math/tex">^\circ</script></span>C in <span class="arithmatex"><span class="MathJax_Preview">[{\rm Na}^{+}]= 0.5</span><script type="math/tex">[{\rm Na}^{+}]= 0.5</script></span> M and <span class="arithmatex"><span class="MathJax_Preview">[{\rm Mg}^{++}]= 0.01</span><script type="math/tex">[{\rm Mg}^{++}]= 0.01</script></span> M:</li>
</ul>
<p><div class="highlight"><pre><span></span><code><span class="n">model2</span> <span class="o">=</span> <span class="n">Model</span><span class="p">(</span><span class="n">material</span><span class="o">=</span><span class="s1">&#39;dna&#39;</span><span class="p">,</span> <span class="n">celsius</span><span class="o">=</span><span class="mi">23</span><span class="p">,</span> <span class="n">sodium</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">magnesium</span><span class="o">=</span><span class="mf">0.01</span><span class="p">)</span>
</code></pre></div>
Note that <code>ensemble</code> is unspecified so it defaults to <code>ensemble='stacking'</code>.</p>
<ul>
<li>Define a model using custom parameters at 45 <span class="arithmatex"><span class="MathJax_Preview">^\circ</span><script type="math/tex">^\circ</script></span>C without coaxial and dangle stacking:</li>
</ul>
<div class="highlight"><pre><span></span><code><span class="n">model3</span> <span class="o">=</span> <span class="n">Model</span><span class="p">(</span><span class="n">material</span><span class="o">=</span><span class="s1">&#39;path/to/my/custom-parameters.json&#39;</span><span class="p">,</span>
    <span class="n">ensemble</span><span class="o">=</span><span class="s1">&#39;nostacking&#39;</span><span class="p">,</span> <span class="n">celsius</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>
</code></pre></div>
</div>
<h3 id="historical-options">Historical options<a class="headerlink" href="#historical-options" title="Permanent link">&para;</a></h3>
<p>For backwards compatibility with NUPACK 3, the following historical complex ensembles without coaxial stacking and with approximate dangle stacking are supported:</p>
<ul>
<li>
<p><code>none-nupack3</code>
No dangle stacking and no coaxial stacking (dangles <code>none</code> option for NUPACK 3)</p>
</li>
<li>
<p><code>some-nupack3</code>
Some dangle stacking and no coaxial stacking (dangles <code>some</code> option for NUPACK 3). A dangle energy is incorporated for each unpaired base flanking a duplex (a base flanking two duplexes contributes only the minimum of the two possible dangle energies).</p>
</li>
<li>
<p><code>all-nupack3</code>
All dangle stacking and no coaxial stacking (dangles <code>all</code> option for NUPACK 3). A dangle energy is incorporated for each unpaired base flanking a duplex (a base flanking two duplexes contributes both possible dangle energies).</p>
</li>
</ul>
<p>For these historical ensembles, base pairs are either Watson-Crick pairs (<code>G</code><span class="arithmatex"><span class="MathJax_Preview">\cdot</span><script type="math/tex">\cdot</script></span><code>C</code> and <code>A</code><span class="arithmatex"><span class="MathJax_Preview">\cdot</span><script type="math/tex">\cdot</script></span><code>U</code> for RNA; <code>G</code><span class="arithmatex"><span class="MathJax_Preview">\cdot</span><script type="math/tex">\cdot</script></span><code>C</code> and <code>A</code><span class="arithmatex"><span class="MathJax_Preview">\cdot</span><script type="math/tex">\cdot</script></span><code>T</code> for DNA) or wobble pairs (<code>G</code><span class="arithmatex"><span class="MathJax_Preview">\cdot</span><script type="math/tex">\cdot</script></span><code>U</code> for RNA; <code>G</code><span class="arithmatex"><span class="MathJax_Preview">\cdot</span><script type="math/tex">\cdot</script></span><code>T</code> for DNA). Note that for the historical ensembles, <code>G</code><span class="arithmatex"><span class="MathJax_Preview">\cdot</span><script type="math/tex">\cdot</script></span><code>T</code> is classified as a DNA wobble pair and not as a mismatch. The historical ensembles prohibit a wobble pair (<code>G</code><span class="arithmatex"><span class="MathJax_Preview">\cdot</span><script type="math/tex">\cdot</script></span><code>U</code> or <code>G</code><span class="arithmatex"><span class="MathJax_Preview">\cdot</span><script type="math/tex">\cdot</script></span><code>T</code>) as a terminal base pair in an exterior loop or a multiloop. As a result, an attempt to evaluate a free energy for a sequence <span class="arithmatex"><span class="MathJax_Preview">\phi</span><script type="math/tex">\phi</script></span> and secondary structure <span class="arithmatex"><span class="MathJax_Preview">s</span><script type="math/tex">s</script></span> that place a wobble pair as a terminal base pair in an exterior loop or multiloop will return <span class="arithmatex"><span class="MathJax_Preview">\overline{\Delta G}(\phi,s)=\Delta G(\phi,s) = \infty</span><script type="math/tex">\overline{\Delta G}(\phi,s)=\Delta G(\phi,s) = \infty</script></span>. These historical ensembles can be used for calculations in combination with the following historical DNA and RNA parameter sets:</p>
<ul>
<li>
<p><code>rna95-nupack3</code>
Same as <code>rna95</code> except that terminal mismatch free energies in exterior loops and multiloops are replaced by two dangle stacking free energies.</p>
</li>
<li>
<p><code>dna04-nupack3</code>
Same as <code>dna04</code> except that G<span class="arithmatex"><span class="MathJax_Preview">\cdot</span><script type="math/tex">\cdot</script></span>T was treated as a wobble pair (analogous to a <code>G</code><span class="arithmatex"><span class="MathJax_Preview">\cdot</span><script type="math/tex">\cdot</script></span><code>U</code> RNA wobble pair) instead of classifying <code>G</code> and <code>T</code> as a mismatch. Note that while terminal mismatch free energies in exterior loops and multiloops are replaced by two dangle stacking free energies, this is the same treatment as in <code>dna04</code>, as terminal mismatch parameters are not public for DNA [<a class="citation" href="#ref-SantaLucia04" id="cite-SantaLucia04">SantaLucia04</a>].</p>
</li>
<li>
<p><code>rna99-nupack3</code>
Parameters from [<a class="citation" href="#ref-Mathews99" id="cite-Mathews99">Mathews99</a>] with terminal mismatch free energies in exterior loops and multiloops replaced by two dangle stacking free energies. Parameters are provided only for 37 <span class="arithmatex"><span class="MathJax_Preview">^\circ</span><script type="math/tex">^\circ</script></span>C.</p>
</li>
</ul>
<h2 id="compute-loop-free-energy">Compute loop free energy<a class="headerlink" href="#compute-loop-free-energy" title="Permanent link">&para;</a></h2>
<p>The <code>loop_energy</code> method operates on a <code>Model</code> object to calculate the <a href="../definitions/#loop-free-energies">loop free energy</a> in kcal/mol. The loop sequence is specified with keyword <code>loop</code> and the loop structure is specified with keyword <code>structure</code>. For example:</p>
<div class="highlight"><pre><span></span><code><span class="n">my_model</span> <span class="o">=</span> <span class="n">Model</span><span class="p">(</span><span class="n">material</span><span class="o">=</span><span class="s1">&#39;RNA&#39;</span><span class="p">,</span> <span class="n">ensemble</span><span class="o">=</span><span class="s1">&#39;stacking&#39;</span><span class="p">)</span>

<span class="c1">#Calculate the free energy of an unstructured strand</span>
<span class="n">dGloop2</span> <span class="o">=</span> <span class="n">my_model</span><span class="o">.</span><span class="n">loop_energy</span><span class="p">(</span><span class="n">loop</span><span class="o">=</span><span class="s1">&#39;AAUU&#39;</span><span class="p">,</span> <span class="n">structure</span><span class="o">=</span><span class="s1">&#39;....&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">dGloop2</span><span class="p">)</span>
<span class="c1"># --&gt; 0.0</span>

<span class="c1">#Calculate the free energy of a hairpin loop</span>
<span class="n">dGloop3</span> <span class="o">=</span> <span class="n">my_model</span><span class="o">.</span><span class="n">loop_energy</span><span class="p">(</span><span class="n">loop</span><span class="o">=</span><span class="s1">&#39;AACCCUU&#39;</span><span class="p">,</span> <span class="n">structure</span><span class="o">=</span><span class="s1">&#39;(.....)&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">dGloop3</span><span class="p">)</span>
<span class="c1"># --&gt; 5.15</span>

<span class="c1">#Calculate the free energy of an exterior loop</span>
<span class="n">dGloop4</span> <span class="o">=</span> <span class="n">my_model</span><span class="o">.</span><span class="n">loop_energy</span><span class="p">(</span><span class="n">loop</span><span class="o">=</span><span class="s1">&#39;AA+UU&#39;</span><span class="p">,</span> <span class="n">structure</span><span class="o">=</span><span class="s1">&#39;((+))&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">dGloop4</span><span class="p">)</span>
<span class="c1"># --&gt; -0.9</span>

<span class="c1">#Calculate the free energy of a multiloop</span>
<span class="n">dGloop5</span> <span class="o">=</span> <span class="n">my_model</span><span class="o">.</span><span class="n">loop_energy</span><span class="p">(</span><span class="n">loop</span><span class="o">=</span><span class="s1">&#39;AAU+ACU+AGU&#39;</span><span class="p">,</span> <span class="n">structure</span><span class="o">=</span><span class="s1">&#39;(.(+).(+).)&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">dGloop5</span><span class="p">)</span>
<span class="c1"># --&gt; 9.355</span>
</code></pre></div>
<h2 id="compute-stacking-state-free-energies">Compute stacking state free energies<a class="headerlink" href="#compute-stacking-state-free-energies" title="Permanent link">&para;</a></h2>
<p>The <code>stack_energies</code> method operates on a <code>Model</code> object to calculate the <a href="../definitions/#loop-free-energies">stacking state free energies</a> for the subensemble of stacking states in a single loop. The loop sequence is specified with keyword <code>loop</code> and the loop structure is specified with keyword <code>structure</code>. The algorithm returns a list of stacking states and the free energy for each in kcal/mol.</p>
<p>For a loop defined as a list of N snippets, a stacking state is specified as a string composed of one letter per snippet. For each snippet, the returned letter is:</p>
<ul>
<li><code>'s'</code> if the snippet contains only 2 nucleotides, each base-paired to a nucleotide in the adjacent snippet, with the two base pairs coaxially stacked on each other</li>
<li><code>'b'</code> if  both the 5<span class="arithmatex"><span class="MathJax_Preview">'</span><script type="math/tex">'</script></span> and 3<span class="arithmatex"><span class="MathJax_Preview">'</span><script type="math/tex">'</script></span> unpaired nucleotides are dangle stacking on adjacent base pairs</li>
<li><code>'5'</code> if only the 5<span class="arithmatex"><span class="MathJax_Preview">'</span><script type="math/tex">'</script></span>-most unpaired base is dangle stacking on its adjacent base pair</li>
<li><code>'3'</code> if only the 3<span class="arithmatex"><span class="MathJax_Preview">'</span><script type="math/tex">'</script></span>-most unpaired base is dangle stacking its adjacent base pair</li>
<li><code>'n'</code> if none of the above apply (i.e., the snippet does not have a dangle at either the 5<span class="arithmatex"><span class="MathJax_Preview">'</span><script type="math/tex">'</script></span> or 3<span class="arithmatex"><span class="MathJax_Preview">'</span><script type="math/tex">'</script></span> end nor does it contain only 2 adjacent nucleotides participating in a coaxial stack)</li>
</ul>
<p>For example, the following figures illustrate snippet annotations for coaxial and dangle stacking states in representative multiloops and exterior loops:</p>
<blockquote>
<p><img src="../figs/stackingnotation.png" alt="Stacking notation" title="Stacking notation" width="600" /></p>
</blockquote>
<p>For a specified multiloop or exterior loop sequence and structure, the <code>stack_energies</code> method returns a set of stacking state strings each with a corresponding stacking state free energy (kcal/mol):</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Calculate the dangle stacking state free energies for an exterior loop</span>
<span class="n">my_model</span><span class="o">.</span><span class="n">stack_energies</span><span class="p">(</span><span class="n">loop</span><span class="o">=</span><span class="s1">&#39;CA+UC&#39;</span><span class="p">,</span> <span class="n">structure</span><span class="o">=</span><span class="s1">&#39;.(+).&#39;</span><span class="p">)</span>
<span class="c1"># --&gt; {&#39;35&#39;: -0.15, &#39;3n&#39;: 0.15, &#39;n5&#39;: 0.35, &#39;nn&#39;: 0.45}</span>

<span class="c1"># Calculate the coaxial stacking state free energies for an exterior loop</span>
<span class="n">my_model</span><span class="o">.</span><span class="n">stack_energies</span><span class="p">(</span><span class="n">loop</span><span class="o">=</span><span class="s1">&#39;AA+U+U&#39;</span><span class="p">,</span> <span class="n">structure</span><span class="o">=</span><span class="s1">&#39;((+)+)&#39;</span><span class="p">)</span>
<span class="c1"># --&gt; {&#39;nnn&#39;: 0.9, &#39;snn&#39;: 0.0}</span>

<span class="c1"># Calculate the coxial stacking state free energies for a multiloop</span>
<span class="n">my_model</span><span class="o">.</span><span class="n">stack_energies</span><span class="p">(</span><span class="n">loop</span><span class="o">=</span><span class="s1">&#39;AU+AU+AU&#39;</span><span class="p">,</span> <span class="n">structure</span><span class="o">=</span><span class="s1">&#39;((+)(+))&#39;</span><span class="p">)</span>
<span class="c1"># --&gt; {&#39;nnn&#39;: 11.9725, &#39;nns&#39;: 10.8725, &#39;nsn&#39;: 10.8725, &#39;snn&#39;: 10.8725}</span>
</code></pre></div>
<p>For loops that are not multiloops or exterior loops, the loop free energy is returned with a string indicating that there is no stacking state. For example, for a hairpin loop:</p>
<div class="highlight"><pre><span></span><code><span class="n">my_model</span><span class="o">.</span><span class="n">stack_energies</span><span class="p">(</span><span class="n">loop</span><span class="o">=</span><span class="s1">&#39;AAAAU&#39;</span><span class="p">,</span> <span class="n">structure</span><span class="o">=</span><span class="s1">&#39;(...)&#39;</span><span class="p">)</span>
<span class="c1"># --&gt; {&#39;n&#39;: 5.85}</span>
</code></pre></div>
<div class="references">
<table>
<tbody>
<tr id="ref-Mathews99">
<td>Mathews99</td>
<td><p>Mathews D.H., Sabina J., Zuker M., Turner D.H.: <i>Expanded sequence dependence of thermodynamic parameters improves prediction of RNA secondary structure</i>. J. Mol. Biol.. <b>288</b>, (1999)</p></td>
</tr>
<tr id="ref-Lu06">
<td>Lu06</td>
<td><p>Lu Z.J., Turner D.H., Mathews D.H.: <i>A set of nearest neighbor parameters for predicting the enthalpy change of RNA secondary structure formation</i>. Nucleic acids research. <b>34</b>, (2006)</p></td>
</tr>
<tr id="ref-Xia98">
<td>Xia98</td>
<td><p>Xia T., SantaLucia J., Burkard M., Kierzek R., Schroeder S., Jiao X., Cox C., Turner D.: <i>thermodynamic parameters for an expanded nearest-neighbor model for formation of RNA duplexes with Watson-Crick base pairs</i>. Biochemistry. <b>37</b>, (1998)</p></td>
</tr>
<tr id="ref-Zuker03">
<td>Zuker03</td>
<td><p>Zuker M.: <i>Mfold web server for nucleic acid folding and hybridization prediction</i>. Nucleic Acids Res.. <b>31</b>, (2003)</p></td>
</tr>
<tr id="ref-Turner10">
<td>Turner10</td>
<td><p>Turner D.H., Mathews D.H.: <i>NNDB: the nearest neighbor parameter database for predicting stability of nucleic acid secondary structure</i>. Nucleic Acids Res.. <b>38</b>, (2010)</p></td>
</tr>
<tr id="ref-Serra95">
<td>Serra95</td>
<td><p>Serra M.J., Turner D.H.: <i>Predicting thermodynamic properties of RNA</i>. Methods Enzymol.. <b>259</b>, (1995)</p></td>
</tr>
<tr id="ref-Allawi97">
<td>Allawi97</td>
<td><p>Allawi H.T., SantaLucia J.: <i>Thermodynamics and NMR of internal G.T mismatches in DNA</i>. Biochemistry. <b>36</b>, (1997)</p></td>
</tr>
<tr id="ref-Allawi98a">
<td>Allawi98a</td>
<td><p>Allawi H.T., SantaLucia J.: <i>Thermodynamics of internal C.T mismatches in DNA</i>. Nucleic Acids Res.. <b>26</b>, (1998)</p></td>
</tr>
<tr id="ref-Allawi98b">
<td>Allawi98b</td>
<td><p>Allawi H.T., SantaLucia J.: <i>Nearest neighbor thermodynamic parameters for internal G.A mismatches in DNA</i>. Biochemistry. <b>37</b>, (1998)</p></td>
</tr>
<tr id="ref-Allawi98c">
<td>Allawi98c</td>
<td><p>Allawi H.T., SantaLucia J.: <i>Nearest-neighbor thermodynamics of internal A.C mismatches in DNA: sequence dependence and pH effects</i>. Biochemistry. <b>37</b>, (1998)</p></td>
</tr>
<tr id="ref-Peyret99">
<td>Peyret99</td>
<td><p>Peyret N., Seneviratne P.A., Allawi H.T., SantaLucia J.: <i>Nearest-neighbor thermodynamics and NMR of DNA sequences with internal A.A, C.C, G.G, and T.T mismatches</i>. Biochemistry. <b>38</b>, (1999)</p></td>
</tr>
<tr id="ref-SantaLucia04">
<td>SantaLucia04</td>
<td><p>SantaLucia J., Hicks D.: <i>The thermodynamics of DNA structural motifs</i>. Annu. Rev. Biophys. Biomol. Struct.. <b>33</b>, (2004)</p></td>
</tr>
<tr id="ref-Mittal24">
<td>Mittal24</td>
<td><p>Mittal A., Turner D.H., Mathews D.H.: <i>NNDB: An expanded database of nearest neighbor parameters for predicting stability of nucleic acid secondary structures</i>. J. Mol. Biol.. <b>436</b>, (2024)</p></td>
</tr>
<tr id="ref-SantaLucia98">
<td>SantaLucia98</td>
<td><p>SantaLucia J.: <i>A unified view of polymer, dumbbell, and oligonucleotide DNA nearest-neighbor thermodynamics</i>. Proc. Natl. Acad. Sci. USA. <b>95</b>, (1998)</p></td>
</tr>
<tr id="ref-Peyret00">
<td>Peyret00</td>
<td><p>Peyret N.: <i>Prediction of nucleic acid hybridization: Parameters and algorithms</i>. (2000)</p></td>
</tr>
<tr id="ref-Bommarito00">
<td>Bommarito00</td>
<td><p>Bommarito S., Peyret N., SantaLucia J.: <i>Thermodynamic parameters for DNA sequences with dangling ends</i>. Nucleic Acids Res.. <b>28</b>, (2000)</p></td>
</tr>
<tr id="ref-Kierzek06">
<td>Kierzek06</td>
<td>Missing citation</td>
</tr>
<tr id="ref-Sugimoto95">
<td>Sugimoto95</td>
<td>Missing citation</td>
</tr>
<tr id="ref-Watkins11">
<td>Watkins11</td>
<td>Missing citation</td>
</tr>
<tr id="ref-Sugimoto00">
<td>Sugimoto00</td>
<td>Missing citation</td>
</tr>
<tr id="ref-Nakano04">
<td>Nakano04</td>
<td>Missing citation</td>
</tr>
<tr id="ref-Nanjundiah25">
<td>Nanjundiah25</td>
<td><p>Nanjundiah A., Fornace M.E., Schulte S.J., Pierce N.A.: <i>Models and algorithms for equilibrium analysis of mixed-material nucleic acid systems</i>. bioRxiv. (2025)</p></td>
</tr>
<tr id="ref-Koehler05">
<td>Koehler05</td>
<td><p>Koehler R.T., Peyret N.: <i>Thermodynamic properties of DNA sequences: Characteristic values for the human genome</i>. Bioinformatics. <b>21</b>, (2005)</p></td>
</tr>
</tbody>
</table>
</div>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../definitions/" class="btn btn-neutral float-left" title="Definitions"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../analysis/" class="btn btn-neutral float-right" title="Analysis Jobs">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="http://nupack.org/" class="fa fa-code-fork" style="color: #fcfcfc"> Nupack</a>
        </span>
    
    
      <span><a href="../definitions/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../analysis/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML"></script>
      <script src="../search/main.js"></script>
      <script src="../js/version-select.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
